src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -3954,47 +3801,416 @@ ColocationGroupDeleteCommand(uint32 colocationId)
+ */
+void
+SendCommandListToSingleWorker(MetadataSyncContext *context, List *commands, int nodeIdx)
 {
-	bool hasColocations = false;
+	/*
+	 * do not send any command to workers if we collect commands.
+	 * Collect commands into metadataSyncContext's collected command
+	 * list.
+	 */
+	if (MetadataSyncCollectsCommands(context))
+	{
+		context->collectedCommands = list_concat(context->collectedCommands, commands);
+		return;
+	}
 
-	StringInfo colocationGroupCreateCommand = makeStringInfo();
-	appendStringInfo(colocationGroupCreateCommand,
-					 "WITH colocation_group_data (colocationid, shardcount, "
-					 "replicationfactor, distributioncolumntype, "
-					 "distributioncolumncollationname, "
-					 "distributioncolumncollationschema)  AS (VALUES ");
+	/* send commands to new workers, the current user should be a superuser */
+	Assert(superuser());
 
-	Relation pgDistColocation = table_open(DistColocationRelationId(), AccessShareLock);
-	Relation colocationIdIndexRel = index_open(DistColocationIndexId(), AccessShareLock);
+	List *workerConnections = context->activatedWorkerConnections;
+	Assert(nodeIdx < list_length(workerConnections));
+	MultiConnection *workerConnection = list_nth(workerConnections, nodeIdx);
 
+	if (context->transactionMode == METADATA_SYNC_TRANSACTIONAL)
+	{
+		SendCommandListToWorkerListWithCoordinatedConnections(list_make1(
+																  workerConnection),
+															  commands);
+	}
+	else
+	{
+		Assert(context->transactionMode == METADATA_SYNC_NON_TRANSACTIONAL);
+		SendCommandListToWorkerListWithBareConnections(list_make1(workerConnection),
+													   commands);
+	}
+}
+
+
+/*
+ * WorkerDropAllShellTablesCommand returns command required to drop shell tables
+ * from workers. When singleTransaction is false, we create transaction per shell
+ * table. Otherwise, we drop all shell tables within single transaction.
+ */
+char *
+WorkerDropAllShellTablesCommand(bool singleTransaction)
+{
+	StringInfo removeAllShellTablesCommand = makeStringInfo();
+	appendStringInfo(removeAllShellTablesCommand, REMOVE_ALL_SHELL_TABLES_COMMAND,
+					 singleTransaction);
+	return removeAllShellTablesCommand->data;
+}
+
+
+/*
+ * PropagateNodeWideObjectsCommandList is called during node activation to
+ * propagate any object that should be propagated for every node. These are
+ * generally not linked to any distributed object but change system wide behaviour.
+ */
+static List *
+PropagateNodeWideObjectsCommandList(void)
+{
+	/* collect all commands */
+	List *ddlCommands = NIL;
+
+	if (EnableAlterRoleSetPropagation)
+	{
+		/*
+		 * Get commands for database and postgres wide settings. Since these settings are not
+		 * linked to any role that can be distributed we need to distribute them seperately
+		 */
+		List *alterRoleSetCommands = GenerateAlterRoleSetCommandForRole(InvalidOid);
+		ddlCommands = list_concat(ddlCommands, alterRoleSetCommands);
+	}
```

Comment:
why did we change this function? 

```C

	if (list_length(ddlCommands) > 0)
	{
		/* if there are command wrap them in enable_ddl_propagation off */
		ddlCommands = lcons(DISABLE_DDL_PROPAGATION, ddlCommands);
		ddlCommands = lappend(ddlCommands, ENABLE_DDL_PROPAGATION);
	}
```

Do we have any other changes in these `CommandList` functions?

Reply:
It is only called by our API. Caller is responsible to add `disable enable ddls`. No other change.